pkg/dogstatsd/parse.go

```diff
@@ -82,6 +85,109 @@ func (p *parser) parseTags(rawTags []byte) []string {
 	return tagsList
 }
 
+func (p *parser) parseMetricSample(message []byte) (dogstatsdMetricSample, error) {
+	// fast path to eliminate most of the gibberish
+	// especially important here since all the unidentified garbage gets
+	// identified as metrics
+	if !hasMetricSampleFormat(message) {
+		return dogstatsdMetricSample{}, fmt.Errorf("invalid dogstatsd message format")
+	}
+
+	rawNameAndValue, message := nextField(message)
+	name, rawValue, err := parseMetricSampleNameAndRawValue(rawNameAndValue)
+	if err != nil {
+		return dogstatsdMetricSample{}, err
+	}
+
+	rawMetricType, message := nextField(message)
+	metricType, err := parseMetricSampleMetricType(rawMetricType)
+	if err != nil {
+		return dogstatsdMetricSample{}, err
+	}
+
+	var setValue []byte
+	var values []float64
+	var value float64
+	if metricType == setType {
+		setValue = rawValue
+	} else {
+		// In case the list contains only one value, dogstatsd 1.0
+		// protocol, we directly parse it as a float64. This avoids
+		// pulling a slice from the float64List and greatly improve
+		// performances.
```

Comment:
How do you know if the float64 has to be returned to the pool later in the process, then?